### ISSUE TITLE
Some Metrics not available on Alder Lake

### ISSUE THREAD

**Issue by s9105947** [tud-zih-energy/lo2s#213]
Dear maintainers,

I tried to perform [Top-Down Microarchitecture Analysis](https://www.intel.com/content/www/us/en/develop/documentation/vtune-cookbook/top/methodologies/top-down-microarchitecture-analysis-method.html) on an Alder Lake processor using lo2s.
However, lo2s does **not find** all required counters.

## Steps to reproduce
Local test system `kallisto`;

```
$ lscpu
[...]
Vendor ID:               GenuineIntel
  Model name:            12th Gen Intel(R) Core(TM) i9-12900K
    CPU family:          6
    Model:               151
[...]
```

Show available counters:

```
$ ls -lh /sys/devices/cpu_core/events/topdown*
-r--r--r-- 1 root root 4.0K Jul  8 13:30 /sys/devices/cpu_core/events/topdown-bad-spec
-r--r--r-- 1 root root 4.0K Jul  8 13:30 /sys/devices/cpu_core/events/topdown-be-bound
-r--r--r-- 1 root root 4.0K Jul  8 13:30 /sys/devices/cpu_core/events/topdown-br-mispredict
-r--r--r-- 1 root root 4.0K Jul  8 13:30 /sys/devices/cpu_core/events/topdown-fe-bound
-r--r--r-- 1 root root 4.0K Jul  8 13:30 /sys/devices/cpu_core/events/topdown-fetch-lat
-r--r--r-- 1 root root 4.0K Jul  8 13:30 /sys/devices/cpu_core/events/topdown-heavy-ops
-r--r--r-- 1 root root 4.0K Jul  8 13:30 /sys/devices/cpu_core/events/topdown-mem-bound
-r--r--r-- 1 root root 4.0K Jul  8 13:30 /sys/devices/cpu_core/events/topdown-retiring
```

```
$ perf list | grep topdown
  topdown-bad-spec OR cpu_atom/topdown-bad-spec/     [Kernel PMU event]
  topdown-bad-spec OR cpu_core/topdown-bad-spec/     [Kernel PMU event]
  topdown-be-bound OR cpu_atom/topdown-be-bound/     [Kernel PMU event]
  topdown-be-bound OR cpu_core/topdown-be-bound/     [Kernel PMU event]
  topdown-br-mispredict OR cpu_core/topdown-br-mispredict/ [Kernel PMU event]
  topdown-fe-bound OR cpu_atom/topdown-fe-bound/     [Kernel PMU event]
  topdown-fe-bound OR cpu_core/topdown-fe-bound/     [Kernel PMU event]
  topdown-fetch-lat OR cpu_core/topdown-fetch-lat/   [Kernel PMU event]
  topdown-heavy-ops OR cpu_core/topdown-heavy-ops/   [Kernel PMU event]
  topdown-mem-bound OR cpu_core/topdown-mem-bound/   [Kernel PMU event]
  topdown-retiring OR cpu_atom/topdown-retiring/     [Kernel PMU event]
  topdown-retiring OR cpu_core/topdown-retiring/     [Kernel PMU event]
```

Now list all counters found by lo2s with `lo2s --list-events | grep topdown`

## Expected Result
All counters above are displayed. (8 for cpu_core, 4 for cpu_atom)

## Actual Result
Only the cpu_atom (E-Core) counters are found

```
$ lo2s --list-events | grep topdown
  cpu_atom/topdown-bad-spec/ *
  cpu_atom/topdown-be-bound/ *
  cpu_atom/topdown-fe-bound/ *
  cpu_atom/topdown-retiring/ *
```

## Additional Notes
This might be caused by the hybrid architecture of Alder Lake, having "P-cores" and "E-cores".
Hence, the sysfs file which typically reside in `/sys/devices/cpu` are split into `/sys/devices/cpu_core` (P-cores) and `/sys/devices/cpu_atom` (E-cores).
Perhaps lo2s does not find all required files? (Notably other cpu_core events are found, e.g. `cpu_core/slots/`)

**Comment by cvonelm** [comment]
In the branch issue-213-alder-lake-metrics I have simply disabled all checks for event availability for now. Could you check that with the right combination of metric leader and metric event you are able to measure the topdown events?

**Comment by cvonelm** [comment]
The problem at hand is, that lo2s checks event availability at start-up by trying to open the event on core 0. This core is most probably one of the low-power Atom Cores, which explains why only the cpu_atom and not the cpu_core events are available.

There are two big solutions to that:

## Introduce --topdown

This is how perf does it. --topdown sets up measurement for all of the topdown events (which also require a custom metric leader) and also calculates topdown metrics which are only available by combining the values of other hardware metrics.

Pro

- easiest to use, doesn't require knowledge about the correct metric leader and how to combine the hardware metrics to get additional metrics, which is otherwise only available buried deep in some Intel Manuals

Cons

- A lot of work only for Alder Lake
- The "get additional metrics by combining hardware metrics" stuff is best suitable for post-processing using Vampir and otf2-python in the lo2s world.

## Introduce --no-checks

Basically, introduce a "I know what im doing" button, that disables most of lo2s handholding and config-checks, for those few situations like here, where the user is smarter than lo2s.

This is by far not as elegant as the solution above (and might not be a solution at all, if testing by @s9105947  shows that there is more stopping us from being able to read topdown events than simply false-negative availability checks), but it is easy to implement and might work outside of this Alder Lake situation, too.

(This is my preferred option)

**Comment by cvonelm** [comment]
This issue is from now on only concerned with the how and why of topdown metrics in lo2s (with a focus on alder lake), alder lake support i.,g. has been moved to another issue.

The two possibilities to explore here are (besides not implementing at all due to cost):

- A dedicated topdown option, like perf.
- possibly minor fixes that allow users to manually select all the topdown metrics.

**Comment by cvonelm** [comment]
I have looked at the example code for alder lake and the documentation at `tools/perf/Documentation/topdown.txt` and it is practically identical to how you can read topdown events Ice Lake.

There seem to be two kind of ways how topdown can be accessed:

## Fixed events (available skylake and up)

In this access mode a normal perf event exists for every topdown metric like `cpu/topdown-fetch-bubbles`.

Support for this is available on every architecture from Skylake up, including Alder Lake. (However Alder Lake itself suffers from lo2s not being hybrid-aware currently)

## Generic event using rdpmc (Available Ice Lake and up)

This is what the topdown code from your Alder Lake example and `tools/perf/Documentation/topdown.txt` does.

With Ice Lake, a new way of reading topdown metrics came to be, which utilizes the rdpmc instruction from userspace. 

The documentation states, that this is faster, but on the other hand it only works on Intel CPUs Ice Lake up.
